data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java

```diff
@@ -22,17 +47,155 @@
+                    multithreadedConsumer = new MultithreadedConsumer(consumerId,
+                            consumerGroupID, consumerProperties, sourceConfig, buffer, pluginMetrics);
+                    executorService.submit(multithreadedConsumer);
+                });
+            } catch (Exception e) {
+                LOG.error("Failed to setup the Kafka Source Plugin.", e);
+                throw new RuntimeException();
+            }
+        });
+    }
+
+    @Override
+    public void stop() {
+        LOG.info("Shutting down Consumers...");
+        executorService.shutdown();
+        try {
+            if (!executorService.awaitTermination(
+                    calculateLongestThreadWaitingTime(), TimeUnit.MILLISECONDS)) {
+                LOG.info("Consumer threads are waiting for shutting down...");
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            if (e.getCause() instanceof InterruptedException) {
+                LOG.error("Interrupted during consumer shutdown, exiting uncleanly...");
+                executorService.shutdownNow();
+                Thread.currentThread().interrupt();
+            }
+        }
+        LOG.info("Consumer shutdown successfully...");
+    }
+
+    private long calculateLongestThreadWaitingTime() {
+        List<TopicsConfig> topicsList = sourceConfig.getTopics();
+        return topicsList.stream().
+                map(
+                        topics -> topics.getThreadWaitingTime().toSeconds()
+                ).
+                max(Comparator.comparingLong(time -> time)).
+                orElse(1L);
+    }
+
+    private Properties getConsumerProperties(TopicsConfig topicConfig) {
+        String schemaType = "";
+        Properties properties = new Properties();
+        properties.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG,
+                topicConfig.getAutoCommitInterval().toSecondsPart());
+        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
+                topicConfig.getAutoOffsetReset());
+        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, sourceConfig.getBootStrapServers());
+        properties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG,
+                topicConfig.getAutoCommit());
+        properties.put(ConsumerConfig.GROUP_ID_CONFIG, topicConfig.getGroupId());
+        schemaType = getSchemaType(sourceConfig.getSchemaConfig().getRegistryURL(), topicConfig.getName(), sourceConfig.getSchemaConfig().getVersion());
+        if (schemaType.isEmpty()) {
+            schemaType = MessageFormat.PLAINTEXT.toString();
+        }
+        setPropertiesForSchemaType(properties, schemaType);
+        if (sourceConfig.getAuthType().equalsIgnoreCase(AuthenticationType.PLAINTEXT.toString())) {
+            setPropertiesForAuth(properties);
+        }
+        return properties;
```

Comment:
Should we add a log properties before starting consumer? It'll be good to know what properties is consumer started with

Reply:
Logger added.